File: 2020/Day_01/aoc_2020_day01.py
from math import prod
from pathlib import Path

import more_itertools as mi


def find_sum_combo(
    entries: list[int], target: int = 2020, n_entries: int = 2
) -> tuple[int, tuple[int]]:
    """
    Iterate through the provided entries & identify the grouping that sums to the target value.

    The product of the grouping is returned, along with the entry group.
    """
    for group in mi.distinct_combinations(entries, n_entries):
        if sum(group) == target:
            return prod(group), group


if __name__ == "__main__":
    puzzle_input_file = Path("./puzzle_input.txt")
    puzzle_input = [int(line) for line in puzzle_input_file.read_text().splitlines()]

    expense_product_pair, entry_pair = find_sum_combo(puzzle_input)
    print(f"Part One: {expense_product_pair}")

    expense_product_trio, entry_trio = find_sum_combo(puzzle_input, n_entries=3)
    print(f"Part Two: {expense_product_trio}")

File: 2020/Day_01/aoc_2020_day01.m
expense_report = parse_expense_report('./puzzle_input.txt');

target_value = 2020;
fprintf('Part One: %u\n', find_entry_product(expense_report, target_value, 2));
fprintf('Part Two: %u\n', find_entry_product(expense_report, target_value, 3));


function [expense_report] = parse_expense_report(filepath)
% Parse the provided expense report file into an nx1 array of integers
    arguments
        filepath char {mustBeFile}
    end

fid = fopen(filepath, 'r');
expense_report = fscanf(fid, '%lu');  % Use 64-bit to keep values as integers vs. double
fclose(fid);
end


function [result_prod] = find_entry_product(expense_report, target, n_values)
% Iterate over all combinations of N_VALUES expense values from the provided expense report
% until a combination is found that sums to the provided TARGET value, then return the product
% of the value combination.
    arguments
        expense_report (1, :)
        target {mustBePositive}
        n_values {mustBePositive}
    end

check_combinations = nchoosek(expense_report, n_values);
sums = sum(check_combinations, 2);  % Row-wise summation

for ii = 1:length(sums)
    if sums(ii) == target
        break
    end
end
result_prod = prod(check_combinations(ii, :));
end
